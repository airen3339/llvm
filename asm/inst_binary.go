package asm

import (
	"fmt"

	"github.com/llir/llvm/ir"
	"github.com/llir/ll/ast"
	"github.com/pkg/errors"
)

// --- [ Binary instructions ] -------------------------------------------------

// ~~~ [ add ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstAdd(inst ir.Instruction, old *ast.AddInst) (*ir.InstAdd, error) {
	i, ok := inst.(*ir.InstAdd)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstAdd, got %T", inst))
	}
	// Overflow flags.
	i.OverflowFlags = irOverflowFlags(old.OverflowFlags())
	// X operand.
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(x.Type(), old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ fadd ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFAdd(inst ir.Instruction, old *ast.FAddInst) (*ir.InstFAdd, error) {
	i, ok := inst.(*ir.InstFAdd)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFAdd, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// X operand.
	// TODO: remove xType in favour of x.Type().
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ sub ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstSub(inst ir.Instruction, old *ast.SubInst) (*ir.InstSub, error) {
	i, ok := inst.(*ir.InstSub)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstSub, got %T", inst))
	}
	// Overflow flags.
	i.OverflowFlags = irOverflowFlags(old.OverflowFlags())
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ fsub ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFSub(inst ir.Instruction, old *ast.FSubInst) (*ir.InstFSub, error) {
	i, ok := inst.(*ir.InstFSub)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFSub, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ mul ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstMul(inst ir.Instruction, old *ast.MulInst) (*ir.InstMul, error) {
	i, ok := inst.(*ir.InstMul)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstMul, got %T", inst))
	}
	// Overflow flags.
	i.OverflowFlags = irOverflowFlags(old.OverflowFlags())
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ fmul ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFMul(inst ir.Instruction, old *ast.FMulInst) (*ir.InstFMul, error) {
	i, ok := inst.(*ir.InstFMul)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFMul, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ udiv ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstUDiv(inst ir.Instruction, old *ast.UDivInst) (*ir.InstUDiv, error) {
	i, ok := inst.(*ir.InstUDiv)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstUDiv, got %T", inst))
	}
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ sdiv ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstSDiv(inst ir.Instruction, old *ast.SDivInst) (*ir.InstSDiv, error) {
	i, ok := inst.(*ir.InstSDiv)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstSDiv, got %T", inst))
	}
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ fdiv ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFDiv(inst ir.Instruction, old *ast.FDivInst) (*ir.InstFDiv, error) {
	i, ok := inst.(*ir.InstFDiv)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFDiv, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ urem ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstURem(inst ir.Instruction, old *ast.URemInst) (*ir.InstURem, error) {
	i, ok := inst.(*ir.InstURem)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstURem, got %T", inst))
	}
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ srem ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstSRem(inst ir.Instruction, old *ast.SRemInst) (*ir.InstSRem, error) {
	i, ok := inst.(*ir.InstSRem)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstSRem, got %T", inst))
	}
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}

// ~~~ [ frem ] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

func (fgen *funcGen) astToIRInstFRem(inst ir.Instruction, old *ast.FRemInst) (*ir.InstFRem, error) {
	i, ok := inst.(*ir.InstFRem)
	if !ok {
		// NOTE: panic since this would indicate a bug in the implementation.
		panic(fmt.Errorf("invalid IR instruction for AST instruction; expected *ir.InstFRem, got %T", inst))
	}
	// Fast math flags.
	i.FastMathFlags = irFastMathFlags(old.FastMathFlags())
	// X operand.
	xType, err := fgen.gen.irType(old.X().Typ())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	x, err := fgen.astToIRTypeValue(old.X())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.X = x
	// Y operand.
	y, err := fgen.astToIRValue(xType, old.Y())
	if err != nil {
		return nil, errors.WithStack(err)
	}
	i.Y = y
	return i, nil
}
