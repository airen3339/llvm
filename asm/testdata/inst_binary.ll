define void @f() {
	add i32 1, 2
	fadd i32 3, 4
	sub i32 5, 6
	fsub i32 7, 8
	mul i32 9, 10
	fmul i32 11, 12
	udiv i32 13, 14
	sdiv i32 15, 16
	fdiv i32 17, 18
	urem i32 19, 20
	srem i32 21, 22
	frem i32 23, 24
	ret void
}
