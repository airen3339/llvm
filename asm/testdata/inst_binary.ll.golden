define i32 @add_1() {
; <label>:0
	%result = add i32 30, 12
	ret i32 %result
}

define <2 x i32> @add_2() {
; <label>:0
	%result = add <2 x i32> <i32 30, i32 12>, <i32 12, i32 30>
	ret <2 x i32> %result
}

define i32 @add_3() {
; <label>:0
	%result = add i32 30, 12
	ret i32 %result
}

define i32 @add_4() {
; <label>:0
	%result = add i32 30, 12, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @add_5() {
; <label>:0
	%result = add i32 30, 12, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @fadd_1() {
; <label>:0
	%result = fadd double 30.0, 12.0
	ret double %result
}

define <2 x double> @fadd_2() {
; <label>:0
	%result = fadd <2 x double> <double 30.0, double 12.0>, <double 12.0, double 30.0>
	ret <2 x double> %result
}

define double @fadd_3() {
; <label>:0
	%result = fadd double 30.0, 12.0
	ret double %result
}

define double @fadd_4() {
; <label>:0
	%result = fadd double 30.0, 12.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define double @fadd_5() {
; <label>:0
	%result = fadd double 30.0, 12.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define i32 @sub_1() {
; <label>:0
	%result = sub i32 50, 8
	ret i32 %result
}

define <2 x i32> @sub_2() {
; <label>:0
	%result = sub <2 x i32> <i32 50, i32 -8>, <i32 8, i32 -50>
	ret <2 x i32> %result
}

define i32 @sub_3() {
; <label>:0
	%result = sub i32 50, 8
	ret i32 %result
}

define i32 @sub_4() {
; <label>:0
	%result = sub i32 50, 8, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @sub_5() {
; <label>:0
	%result = sub i32 50, 8, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @fsub_1() {
; <label>:0
	%result = fsub double 50.0, 8.0
	ret double %result
}

define <2 x double> @fsub_2() {
; <label>:0
	%result = fsub <2 x double> <double 50.0, double -8.0>, <double 8.0, double -50.0>
	ret <2 x double> %result
}

define double @fsub_3() {
; <label>:0
	%result = fsub double 50.0, 8.0
	ret double %result
}

define double @fsub_4() {
; <label>:0
	%result = fsub double 50.0, 8.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define double @fsub_5() {
; <label>:0
	%result = fsub double 50.0, 8.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define i32 @mul_1() {
; <label>:0
	%result = mul i32 21, 2
	ret i32 %result
}

define <2 x i32> @mul_2() {
; <label>:0
	%result = mul <2 x i32> <i32 21, i32 2>, <i32 2, i32 21>
	ret <2 x i32> %result
}

define i32 @mul_3() {
; <label>:0
	%result = mul i32 21, 2
	ret i32 %result
}

define i32 @mul_4() {
; <label>:0
	%result = mul i32 21, 2, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @mul_5() {
; <label>:0
	%result = mul i32 21, 2, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @fmul_1() {
; <label>:0
	%result = fmul double 21.0, 2.0
	ret double %result
}

define <2 x double> @fmul_2() {
; <label>:0
	%result = fmul <2 x double> <double 21.0, double 2.0>, <double 2.0, double 21.0>
	ret <2 x double> %result
}

define double @fmul_3() {
; <label>:0
	%result = fmul double 21.0, 2.0
	ret double %result
}

define double @fmul_4() {
; <label>:0
	%result = fmul double 21.0, 2.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define double @fmul_5() {
; <label>:0
	%result = fmul double 21.0, 2.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define i32 @udiv_1() {
; <label>:0
	%result = udiv i32 84, 2
	ret i32 %result
}

define <2 x i32> @udiv_2() {
; <label>:0
	%result = udiv <2 x i32> <i32 84, i32 42>, <i32 2, i32 1>
	ret <2 x i32> %result
}

define i32 @udiv_3() {
; <label>:0
	%result = udiv i32 84, 2
	ret i32 %result
}

define i32 @udiv_4() {
; <label>:0
	%result = udiv i32 84, 2, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @udiv_5() {
; <label>:0
	%result = udiv i32 84, 2, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @sdiv_1() {
; <label>:0
	%result = sdiv i32 -84, -2
	ret i32 %result
}

define <2 x i32> @sdiv_2() {
; <label>:0
	%result = sdiv <2 x i32> <i32 -84, i32 42>, <i32 -2, i32 1>
	ret <2 x i32> %result
}

define i32 @sdiv_3() {
; <label>:0
	%result = sdiv i32 -84, -2
	ret i32 %result
}

define i32 @sdiv_4() {
; <label>:0
	%result = sdiv i32 -84, -2, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @sdiv_5() {
; <label>:0
	%result = sdiv i32 -84, -2, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @fdiv_1() {
; <label>:0
	%result = fdiv double 84.0, 2.0
	ret double %result
}

define <2 x double> @fdiv_2() {
; <label>:0
	%result = fdiv <2 x double> <double 84.0, double 42.0>, <double 2.0, double 1.0>
	ret <2 x double> %result
}

define double @fdiv_3() {
; <label>:0
	%result = fdiv double 84.0, 2.0
	ret double %result
}

define double @fdiv_4() {
; <label>:0
	%result = fdiv double 84.0, 2.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define double @fdiv_5() {
; <label>:0
	%result = fdiv double 84.0, 2.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define i32 @urem_1() {
; <label>:0
	%result = urem i32 85, 43
	ret i32 %result
}

define <2 x i32> @urem_2() {
; <label>:0
	%result = urem <2 x i32> <i32 85, i32 162>, <i32 43, i32 60>
	ret <2 x i32> %result
}

define i32 @urem_3() {
; <label>:0
	%result = urem i32 85, 43, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define i32 @srem_1() {
; <label>:0
	%result = srem i32 85, -43
	ret i32 %result
}

define <2 x i32> @srem_2() {
; <label>:0
	%result = srem <2 x i32> <i32 85, i32 162>, <i32 -43, i32 -60>
	ret <2 x i32> %result
}

define i32 @srem_3() {
; <label>:0
	%result = srem i32 85, -43, !baz !{!"qux"}, !foo !{!"bar"}
	ret i32 %result
}

define double @frem_1() {
; <label>:0
	%result = frem double 85.0, 43.0
	ret double %result
}

define <2 x double> @frem_2() {
; <label>:0
	%result = frem <2 x double> <double 85.0, double 162.0>, <double 43.0, double 60.0>
	ret <2 x double> %result
}

define double @frem_3() {
; <label>:0
	%result = frem double 85.0, 43.0
	ret double %result
}

define double @frem_4() {
; <label>:0
	%result = frem double 85.0, 43.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}

define double @frem_5() {
; <label>:0
	%result = frem double 85.0, 43.0, !baz !{!"qux"}, !foo !{!"bar"}
	ret double %result
}
